### DISCUSSION TITLE
Rotation of Redaction Text

### DISCUSSION THREAD

**Discussion by mwade625** [pymupdf/PyMuPDF#1729] [Looking for help]
When I create redactions, the redaction box is correctly placed over the text, but that is passed in to fill in the redaction box, is written vertically.  My document is show in landscape mode and visually the text runs left to right, but it is actually rotated 90 and the direction of the text is (0.0, -1.0).  

The redaction box is covering the text as expected, but the text that is written into the redaction area is written as if the page is un-rotated.  I have tried to ROTATE the annotation 90 and write the text but it always comes out this way.  My code is as follows:

redaction = page.add_redact_annot(quad=text_coords, text=redaction_text, textcolor=(1.0,1.0,1.0), fill=(0.0,0.0,0.0), cross_out=False)
redaction.set_colors(fill_color, stroke_color)
highlight.set_info(info=redaction_info)
highlight.update()

I noticed in the code for add_redact_anon() that it is always setting the page rotation to 0 so it seems there is NO way to actually get the text to write out properly?   I have tried setting the page rotation to 0 before any of the above calls, but it doesn't help.
Any thoughts on HOW to get the rotation of the text in the redaction box to match what I want?  I have the same problem with regular annotations as well.

**Comment by JorjMcKie** [comment]
Text rotation is not supported with redactions. It is however with freetext annotations (multiples of 90°).

Redaction text, in the end, will become regular text of the page (in contrast to freetext annots).
So you coul use a 2-step approach: add and apply a redaction, then use normal text insertion to put stuff into the now free rectangle.

**Reply by mwade625** [reply]
Thanks, I was somewhat aware that this wasn't supported, but your replacement idea is great and I will try it.  One remaining issue is that regular annotations put the elipses at the end of the annotation rectangle (curved ends) and in this case it puts them at the top and bottom.

Unfortunately, because the text coordinates are going up/down but the image is rotated, not matter how I rotate the page the annotation rectangle will be drawn incorrectly.  Is there anyway to turn off that curved end's?

**Reply by JorjMcKie** [reply]
Curved ends? With PyMuPDF only rectangles with sharp corners are drawn.
Otherwise if you insert text separately, you can choose the writing direction such that it fits the page rotation.

**Comment by mwade625** [comment]
Here is a screen shot of the add_highlight_annot() result.  You can see the curved ends of the annotation area.  It looks this way if I render it with pymupdf to image or use any number of PDF viewers.  

![image](https://user-images.githubusercontent.com/65691688/170744388-a96a4082-1ca6-4f41-a067-35e0b6c7004c.png)


I have a document where the text is displayed going from left to right but the document is actually rotated 90.  If tried adjusting the page rotation first and then apply the annotation, as well as chaning the coordinates but I can't get the annotation to work properly.  

I was able to get around the issue with Redactions and text writing by using your solution of using the freetext annotation for the text.  I just can't figure out how to do a highlight redaction and get it to obey the actual page rotation.

**Comment by JorjMcKie** [comment]
These are **_highlight_** annotations, no redactions!
To make behave **them** correctly, you must not use a simple rectangle as argument, because MuPDF needs to know where is top-left-right-bottom to draw the right thing.
Specify a quad instead. This is documented somewhere, including some previous related issue and an example script in the Utilities repo.

**Reply by mwade625** [reply]
I followed your suggestion and just started doing my own annotations and the hybrid redactions and that code does work in this case.  Thanks for the tips!
